metagraph/src/annotation/binary_matrix/row_diff/row_diff.hpp

```diff
@@ -154,6 +154,8 @@ BinaryMatrix::SetBitPositions RowDiff<BaseMatrix>::get_row(Row row) const {
 
     while (!anchor_[row]) {
         boss_edge = boss.row_diff_successor(boss_edge, rd_succ);
+        if (!boss_edge)
+            return result;
```

Comment:
This must never happen, why did you need to add this?

Reply:
If row ids corresponding to dummy sink k-mers are passed, then `row_diff_successor` crashes. For the other annotation representations, these indices just return empty annotations, so I wanted to have that same behaviour here.

Reply:
Hm, it was designed so that you never call it for not existing/dummy nodes.

Why do you need to do that?

Reply:
There are two cases where this can come up
1) slice_columns (to be renamed call_columns) for now iterates through each row index and checks it. This one is easy to fix since we can check for dummy sinks while iterating.
2) call_outgoing_kmers and adjacent_outgoing_nodes in DBGSuccinct returns dummy sinks, so if these methods are used to populate a row index vector, then it'll include dummy sinks. I don't remember if there's a reason why we kept it like this.

Reply:
Both of these would be incredibly inefficient to call with the current implementation.

I suggest better throw error on those calls and not change this implementation.

Reply:
The `call_outgoing_kmers` approach is used in alignment for traversal (after the alignment is done, all traversed nodes can be queried as a batch). In basically all cases where this can occur, checks can be put in other parts of the code to prevent it from propagating here.

In the end, not fixing this just leaves a conceptual issue: a row index corresponding to a dummy k-mer is still a technically valid input, but it causes this function to crash.

Reply:
Ok, but you never want to align to dummy nodes. So you should check and skip them. Then you can extract annotations for the batch of actual real nodes.

> In the end, not fixing this just leaves a conceptual issue: a row index corresponding to a dummy k-mer is still a technically valid input, but it causes this function to crash.

The dummy nodes are valid for BOSS, but not on the higher abstraction levels (SuccinctDBG/DeBruijnGraph), where they are as invalid as any negative index. So having UB or throwing an error for that looks ok to me, there are no conceptual issues.